### PR TITLE
OOP school library: decorate a class

### PR DESCRIPTION
**Pull request by AdnanOlarmmi**
# OOP school library: decorate a class

Hi Reviewer, 
Patrick Hi

## Learning objectives
- Implement composition (as an example of the design pattern).
- Run a program using the command line.

## Description
In this project, I used the Decorator design pattern to correct the names of people.

### General requirements

- Made sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
- Made sure that I used correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
- Made sure that I documented my work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).

### Ruby requirements
- Followed the list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).

### Project requirements
- Checked the usage of [Decorator pattern in Ruby](https://refactoring.guru/design-patterns/decorator/ruby/example#example-0)
- Thought about how I can use two decorators in order to capitalize and trim people's names.

#### Interface
- Created a class `Nameable`.
   - Implemented a method called `correct_name` that raised a `NotImplementedError`.

#### Turned the Person class to Nameable
- Made sure that the Person class inherits from Nameable
- Made sure that this class has a method `correct_name` implemented. It simply returns the name attribute.

#### Prepared base Decorator
- Made sure that it inherits from Nameable.
- In the constructor, I assigned a nameable object from params to an instance variable.
- Implemented the `correct_name` method that returns the result of the correct_name method of the `@nameable`.

#### Prepared CapitalizeDecorator and TrimmerDecorator
- For the CapitalizeDecorator:
    - Created a class that inherits from the base Decorator class.
    - Implemented a method `correct_name` that capitalizes the output of `@nameable.correct_name`.
- For the TrimmerDecorator:
    - Created a class that inherits from the base Decorator class.
    - Implemented a method `correct_name` that makes sure that the output of `@nameable.correct_name` has a maximum of 10 characters. If it's longer it should trim the word.

### Saw the decorators in action
\Tried the following code and check if you managed to decorate your person:

``` ruby
person = Person.new(22, 'maximilianus')
person.correct_name
capitalized_person = CapitalizeDecorator.new(person)
puts capitalized_person.correct_name
capitalized_trimmed_person = TrimmerDecorator.new(capitalized_person)
puts capitalized_trimmed_person.correct_name
```

and it returned 
```
Maximilianus
Maximilian
```

## Work and submission mode

- Activity submitted **individually.**

## Appreciation
- Thanks for the review, it means a lot.

------

### CODE REVIEW

File: base_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+require_relative 'nameable'
+
+class BaseDecorator < Nameable
+  def initialize(nameable)
+    @nameable = nameable
```
Comment (VICTOR-CHUKWU) [inline base_decorator.rb:7]: - [ ] kindly make nameable accessible with attr_accessor

**Review by VICTOR-CHUKWU — CHANGES_REQUESTED** [review]
Hi @AdnanOlarmmi  👋  

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
- Good PR
- Good commit
- All linters passed


## Required Changes ♻️

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by AdnanOlarmmi** [comment]
Thanks @VICTOR-CHUKWU

**Review by GzimAsani — APPROVED** [review]
# STATUS: APPROVED :100:         
## There is nothing else to say than it's time to merge... :partying_face: 
Cheers and Happy coding! :clap:  :computer:  :wine_glass: 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


## Optional suggestions


_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A
### Please be reminded to not open a new pull request after you get a review, only use the same one :thought_balloon: 


 **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
 
 _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**